### ISSUE TITLE
[bug][framework] transaction effects returned by test_scenario do not include all created objects

### ISSUE THREAD

**Issue by karlwuest** [MystenLabs/sui#12311]
The transaction effects returned by `test_scenario::next_tx` do not include all created objects when they are added to an ObjectBag. This behaviour is inconsistent with the effects reported in the same contract call using the sui client.

## Steps to Reproduce Issue

Run tests for the following example move code:

```
module bag_test::bag_test {
    use sui::object::{Self, UID};
    use sui::transfer;
    use sui::tx_context::TxContext;
    use sui::object_bag::{ObjectBag, Self};

    struct Item has key, store {
        id: UID,
    }

    struct Container has key {
        id: UID,
        items: ObjectBag,
    }

    public entry fun create_container(ctx: &mut TxContext) {
        let id = object::new(ctx);
        let items = object_bag::new(ctx);
        transfer::share_object(Container {
            id,
            items,
        })
    }

    public entry fun create_item(
        container: &mut Container, ctx: &mut TxContext
    ) {
        let id = object::new(ctx);
        let item_id = object::uid_to_inner(&id);
        object_bag::add(
            &mut container.items,
            item_id,
            Item {
                id,
            }
        );
    }

    #[test]
    fun test_bag_example() {
        use sui::test_scenario;
        use std::vector;
        let client_address = @0xc0ffee;
        let scenario_val = test_scenario::begin(client_address);
        let scenario = &mut scenario_val;
        create_container(test_scenario::ctx(scenario));
        test_scenario::next_tx(scenario, client_address);
        let container = test_scenario::take_shared<Container>(scenario);
        create_item(&mut container, test_scenario::ctx(scenario));
        test_scenario::return_shared(container);
        let tx_effects = test_scenario::next_tx(scenario, client_address);
        let created = test_scenario::created(&tx_effects);

        // The `created` vector should have at least one element, namely the id of the created item
        assert!(vector::length(&created) >= 1, 0);

        test_scenario::end(scenario_val);
    }
}
```

## Expected Result

The test should succeed. If the same code is deployed on a local sui testnet, calling `create_item` using the sui client, the effects will include the item in the created objects:

```
----- Transaction Effects ----
Status : Success
Created Objects:
  - ID: 0x487ed14d760bb0ed92d6d5e8eb723a390d92be5455abd87b2026d6e29dbc340a , Owner: Object ID: ( 0x68dca8e9b605efe457bf7af8e695be171b79eea2b18178f8744ab390a63d9d66 )
  - ID: 0x92b34b01c18beec96da077ceb766684eac46f77f80f32fb230e0bd38f81b6efb , Owner: Object ID: ( 0x487ed14d760bb0ed92d6d5e8eb723a390d9
2be5455abd87b2026d6e29dbc340a )
````

where
```
$ sui client object 0x92b34b01c18beec96da077ceb766684eac46f77f80f32fb230e0bd38f81b6efb
----- 0xd22cf812d9ed24a70054c91ce84dfec6303341a60150d21aeae0a13ea83d6870::bag_test::Item (0x92b34b01c18beec96da077ceb766684eac46f77f80f32fb230e0bd38f81b6efb[0x9]) -----
Owner: Object ID: ( 0x487ed14d760bb0ed92d6d5e8eb723a390d92be5455abd87b2026d6e29dbc340a )
Version: 0x9
Storage Rebate: 1284400
Previous Transaction: TransactionDigest(3qUhnyCBVyLpH3amG2BYGPujnbao8X8ghNnnRDjo8WFv)
----- Data -----
type: 0xd22cf812d9ed24a70054c91ce84dfec6303341a60150d21aeae0a13ea83d6870::bag_test::Item
id: 0x92b34b01c18beec96da077ceb766684eac46f77f80f32fb230e0bd38f81b6efb
```
## Actual Result

The test fails. The transaction effects returned by `test_scenario` include no created objects.

## System Information

* OS: macOS Ventura 13.4
* Compiler: cargo 1.69.0
* sui version: Tested with sui at version https://github.com/MystenLabs/sui/commit/9588990ecf2b1c359b1cd2f2afbc3dce8ac110ba and https://github.com/MystenLabs/sui/commit/8b681515c0cf435df2a54198a28ab4ef574d202b

**Comment by rvantonder** [comment]
Hey @karlwuest thanks for the detailed steps to reproduce!

TL;DR: After digging in, this looks like a limitation in `test_scenario`, and I'm not sure there's a fix or improvement we can to `test_scenario`. But I will raise with some folks who know more and then revert here. My suggested workaround is to test for `>= 1` items in the container by:

- Getting the ID of the container in the test (e.g., you can get that from the `shared` effects at the end of the test).

- Retrieving the object by ID from the store and testing length on that.

I.e., instead of checking state that depends on the fidelity of reported effects by `test_scenario`, instead check state based on stored / mutated objects.

--- 

Long version, according to my best understanding after going through the code:

- The transaction effects reported by `test_scenario::next_tx` aren't accumulated. Each invocation of `next_tx` returns the results / transaction effects of the previous transaction.
  - This means the `created` effects for `create_container` are not observable toward the end of test (after calling `next_tx` again), but you will see them in the results of `next_tx` called earlier in the test.

- Transaction effects for `test_scenario` are committed / recorded on `sui::transfer` calls.
  - This means transaction effects are observable for `next_tx` after calling `create_container` because of `transfer::share_object`. You can see a `create` effect on `next_tx` for this call in the debug output below **part 1**.
  - This also means transaction effects are observable for `next_tx` after calling `return_shared`, which indirectly calls `transfer::share_object_impl` (see **part 3** output below)
  - Because of this, there is no observable `create` effect when calling `create_item` (limitation here: there is no recording of this in `test_scenario`, and no function (e.g., like `transfer` is invoked to expose this info for us to record. effectively, there is no transaction). See **part 2** output below.
  - Once `return_shared` is called, we do see that `container` is mutated, and that fact is recorded in `written` effects, but no effects are `created` (working-as-intended, because per bullet above, `test_scenario` doesn't know about this newly created item). See **part 3**.

- From here it makes sense that we can see observable effects when calling functions on a local testnet (effects are accumulated and committed / recorded for the transaction involving child object, which test_scenario doesn't see in the window where `create_item` is called in the test).

So I'm not sure there's a way to improve this based on how `test_scenario works`, but will forward for input from rest of the folks!

Code:

```rs
    #[test]                                                                                                                                                                                                        
    fun test_bag_example() {                                                                                                                                                                                       
        use sui::test_scenario;                                                                                                                                                                                    
        use std::vector;                                                                                                                                                                                           
        let client_address = @0xc0ffee;                                                                                                                                                                            
        let scenario_val = test_scenario::begin(client_address);                                                                                                                                                   
        let scenario = &mut scenario_val;                                                                                                                                                                          
        create_container(test_scenario::ctx(scenario));                                                                                                                                                            
        let tx_effects = test_scenario::next_tx(scenario, client_address);                                                                                                                                         
                                                                                                                                                                                                                   
        sui::test_utils::print(b"after create_container");                                                                                                                                                         
        std::debug::print(&tx_effects);                                                                                                                                                                            
                                                                                                                                                                                                                   
        let container = test_scenario::take_shared<Container>(scenario);                                                                                                                                           
        create_item(&mut container, test_scenario::ctx(scenario));                                                                                                                                                 
                                                                                                                                                                                                                   
        let tx_effects = test_scenario::next_tx(scenario, client_address);                                                                                                                                         
        sui::test_utils::print(b"after create_item");                                                                                                                                                              
        std::debug::print(&tx_effects);                                                                                                                                                                            
        // std::debug::print(&container);  // uncomment to see that item is added                                                                                                                                                                        
                                                                                                                                                                                                                   
        test_scenario::return_shared(container);                                                                                                                                                                   
                                                                                                                                                                                                                   
        let tx_effects = test_scenario::next_tx(scenario, client_address);                                                                                                                                         
        sui::test_utils::print(b"after return_shared");                                                                                                                                                            
        std::debug::print(&tx_effects);                                                                                                                                                                            
        let created = test_scenario::created(&tx_effects);                                                                                                                                                         
                                                                                                                                                                                                                   
        // The `created` vector should have at least one element, namely the id of the created item                                                                                                                
        assert!(vector::length(&created) >= 1, 0);                                                                                                                                                                 
                                                                                                                                                                                                                   
        test_scenario::end(scenario_val);                                                                                                                                                                          
    }
```

Debug log, separated for readability

**part 1** `created` effect for container

```
[debug] "after create_container"
[debug] 0x2::test_scenario::TransactionEffects {
  created: [
    0x2::object::ID {
      bytes: @0x34401905bebdf8c04f3cd5f04f442a39372c8dc321c29edfb4f9cb30b23ab96
    }
  ],
  written: [],
  deleted: [
    0x2::object::ID {
      bytes: @0xa4137e2ed945e7f4e8c307001c32e97472fb3eebb8f7b4ab7534b447d22da89b
    }
  ],
  transferred_to_account: 0x2::vec_map::VecMap<0x2::object::ID, address> {
    contents: []
  },
  transferred_to_object: 0x2::vec_map::VecMap<0x2::object::ID, 0x2::object::ID> {
    contents: []
  },
  shared: [
    0x2::object::ID {
      bytes: @0x34401905bebdf8c04f3cd5f04f442a39372c8dc321c29edfb4f9cb30b23ab96
    }
  ],
  frozen: [],
  num_user_events: 0
}
```

**part 2** no observable effects in `test_scenario` for `create_item`

```
[debug] "after create_item"
[debug] 0x2::test_scenario::TransactionEffects {
  created: [],
  written: [],
  deleted: [],
  transferred_to_account: 0x2::vec_map::VecMap<0x2::object::ID, address> {
    contents: []
  },
  transferred_to_object: 0x2::vec_map::VecMap<0x2::object::ID, 0x2::object::ID> {
    contents: []
  },
  shared: [],
  frozen: [],
  num_user_events: 0
}
```

**part 3** `next_tx` returns effects (sees mutated / `written` container).

```
[debug] "after return_shared"
[debug] 0x2::test_scenario::TransactionEffects {
  created: [],
  written: [
    0x2::object::ID {
      bytes: @0x34401905bebdf8c04f3cd5f04f442a39372c8dc321c29edfb4f9cb30b23ab96
    }
  ],
  deleted: [
    0x2::object::ID {
      bytes: @0xa4137e2ed945e7f4e8c307001c32e97472fb3eebb8f7b4ab7534b447d22da89b
    }
  ],
  transferred_to_account: 0x2::vec_map::VecMap<0x2::object::ID, address> {
    contents: []
  },
  transferred_to_object: 0x2::vec_map::VecMap<0x2::object::ID, 0x2::object::ID> {
    contents: []
  },
  shared: [
    0x2::object::ID {
      bytes: @0x34401905bebdf8c04f3cd5f04f442a39372c8dc321c29edfb4f9cb30b23ab96
    }
  ],
  frozen: [],
  num_user_events: 0
}
```

**Comment by karlwuest** [comment]
Thanks for looking into this!
The `assert!(vector::length(&created) >= 1, 0)` is just in this example to show the bug. My original goal was to retrieve the ID of the created object (which is needed for future function calls). The workaround that we've been using is to have the `create_item` function return the `item_id` of the object. This works for us, but, of course, I still wanted to report the bug.

**Comment by stefan-mysten** [comment]
@rvantonder @karlwuest is this issue still relevant?

**Comment by rvantonder** [comment]
Don't remember the exact details (one sign this is too old to keep open) and since it's not easily fixable (second sign) and since there's a plausible workaround, I think we can close this
